### PR TITLE
[Feature] Package diff 

### LINKED ISSUE #532: Package comparison
Packages -> Remote instance -> Contextual Action: Package comparison: MetadataComparison.tsx

`MetadataComparison.diff(localPackageNotSaved, remotePackage) -> {updated: ..., created: ...}`

- localPackage using IDS from remotePackage.

Front-end. Modal rendering a summary of:
   - [x] Green: added objects
   - [x] Orange: modified objects

Layout:
- [x] New contextual action on package view 'Compare with local instance'
- [x] Open a modal dialog fullscreen
- [x] For now show only difference (no need for two sections). For example
   - 2 New DEs (include name and some relevant fields)
   - 3 Deleted DEs (include name and some relevant fields) -> NOT POSSIBLE
  - 4 Modified Dataset (include name and some relevant fields)
- [x] Two buttons at the bottom: Import / Cancel
- [x] Check if import is already done and we can reuse the method -> ./domain/metadata/usecases/ImportMetadataUseCase.ts
- [ ] Test for getMetadataDiff -> NOT FOR NOW.
- [x] Check performance with orgUnits

Later:
Left-side: remote instance → Right-side: local instance

### PR DESCRIPTION
**Pull request by tokland**
### :pushpin: References

-   **Issue:** Closes #532

### :memo: Implementation

- We create a local (unsaved) package to compare with the remote package. @SferaDev This is the approach we discussed and it works well, only one caveat: let's say a data set is not found locally, it's only on the remote package; the diff will show its dependencies (data elements, sections, ...) as being new (even though they might exist), simply because the data set was not present in the local package, so neither are its dependencies. I don't think it's a major issue, but maybe you can see a workaround.
- As we agreed, we pass the `compositionRoot` to the `DiffPackageUseCase` to DRY the use case.
- Ignored fields: `["created", "lastUpdated", "lastUpdatedBy", "version"]`.
- I've tried a simple approach when comparing objects: values are converted to strings (easy to compare and show). A caveat for arrays: DHIS2 uses an array both for list (sorted) and a set (unsorted). The diff function does not know which one a particular attribute is. Currently, it assumes it's doing a set comparison (which is the most typical), but, if it's a list, it would fail to detect an item reordering. If that's a problem, proposed solution: change d2-api to save the model schema somewhere so we check if it's a list or a set (`klass`: `"java.util.List"` or `"java.util.Set"`).

### :art: Screenshots

image

### CODE REVIEW

File: src/domain/packages/usecases/DiffPackageUseCase.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public async execute(
+        packageId: string,
+        remoteInstance: Instance
+    ): Promise<Either<DiffPackageUseCaseError, MetadataPackageDiff>> {
+        const storageRepository = this.repositoryFactory.get<StorageRepositoryConstructor>(
+            Repositories.StorageRepository,
+            [remoteInstance]
+        );
+
+        const package_ = await storageRepository.getObjectInCollection<Package>(
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/domain/packages/usecases/DiffPackageUseCase.ts:31]: > is it possible to detect and error in the type definition of getObjectInCollection that T is not a class instance

From the outset, the only difference you'd see is that bare interfaces have no functions. With some TS machinery, you could reject a type with functions being passed, but it's quite a hack. 

I think it's better to rethink the `get<SomeType>("someName")` infrastructure.

Temporarily, we could also write: `getObjectInCollection<T extends Type1 | Type2 | ...>`

File: src/presentation/common/components/package-list-table/PackageListTable.tsx [hunk 2]
```diff
@@ -66,6 +73,17 @@ export const PackagesListTable: React.FC<ModuleListPageProps> = ({
         [compositionRoot, remoteInstance, snackbar]
     );
 
+    const openPackageDiffDialog = useCallback(
+        async (ids: string[]) => {
+            const packageId = ids[0];
+            const package_ = rows.find(row => row.id === packageId);
+            if (packageId && package_) setPackageToDiff({ id: packageId, name: package_.name });
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/presentation/common/components/package-list-table/PackageListTable.tsx:80]: It should be usually redundant, but if you had a bug somewhere, `ids` might be empty or the package could not be in rows, then you'd get a cryptic error. It should be `const package_ = packageId ? rows.find(row => row.id === packageId) : null`, to be strict.

**Comment by EyeSeeTeaBotTest** [comment]
<h3>Tests failed with errors :rotating_light:</h3><br />Build commit Merge 4bf685ab44fe750b6e8f462b0f09017ad4649161 into 6b56b7ecfac132c28fb81ad3c99f097faa5985cd with commit id 6a4f57f5ae061b99df8680738b0c8ed38bdfb0dc on branch feature/package-diff-532<br /><br />Travis log available at <a href=https://travis-ci.org/EyeSeeTea/metadata-synchronization/builds/717543499 target=_blank>builds/717543499</a><br /><br />See cypress results at <a href=https://eyeseetea.github.io/feedback-ci/EyeSeeTea/metadata-synchronization/pr/574/6a4f57f5ae061b99df8680738b0c8ed38bdfb0dc target=_blank>feedback-ci/EyeSeeTea/metadata-synchronization/pr/574/6a4f57f5ae061b99df8680738b0c8ed38bdfb0dc</a>

File: src/domain/packages/usecases/DiffPackageUseCase.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SferaDev) [reply src/domain/packages/usecases/DiffPackageUseCase.ts:31]: Yeah the ``get<SomeType>("someName")`` is on the refactor's task and we will revisit it asap. It's okay to leave it as-is for now

File: src/presentation/common/components/package-list-table/PackageListTable.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (tokland) [reply src/presentation/common/components/package-list-table/PackageListTable.tsx:80]: About the snippet: I know you like to de-structure as early as possible. Somehow, for me, it adds one indication level when I look at the variables, and it's less clear what object attributes I am working with. Not in this case, of course, which is very short, but in the general case.

But I do think it's completely fine to de-structure things as you do. 

However, In your snippet, for example, this `?? {}` has made that `name` has type `any`. So you'd have to write `?? {name: undefined}` or something along the line, which is not so nice. And then, on the check, it's not the same to not have an object or to be working with the fallback case. Very nit-picky, yes, but on some cases it's relevant (not here). So I don't have to think on all these details, I tend to use full objects :)

[the remark this answers is not in the document]
Comment (SferaDev) [reply src/presentation/common/components/package-list-table/PackageListTable.tsx:80]: > However, In your snippet, for example, this `?? {}` has made that `name` has type `any`. So you'd have to write `?? {name: undefined}` or something along the line, which is not so nice.

I velieve that happened in old versions of TS, but it doesn't look like that anymore, to me it correctly infers ``string | undefined`` here.

image

File: src/presentation/webapp/components/packages-diff-dialog/PackagesDiffDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import SyncSummary from "../sync-summary/SyncSummary";
+import { usePackageImporter, getTitle, getChange } from "./utils";
+
+export interface PackagesDiffDialogProps {
+    onClose(): void;
+    remoteInstance: Instance;
+    remotePackage: { id: string; name: string };
+}
+
+export type PackageToDiff = { id: string; name: string };
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/presentation/webapp/components/packages-diff-dialog/PackagesDiffDialog.tsx:23]: I think it's completely fine to use `NamedRef`, it's generic and in the structural spirit of JS/TS, a thingy with an id and a name.

File: src/presentation/webapp/components/packages-diff-dialog/utils.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+            setSyncReport(report);
+        }
+
+        import_()
+            .catch(err => snackbar.error(err.message))
+            .finally(() => loading.reset());
+    }, [packageName, metadataDiff, compositionRoot, loading, snackbar, api]);
+
+    return [importPackage, syncReport, closeSyncReport] as const;
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/presentation/webapp/components/packages-diff-dialog/utils.ts:69]: First this hook had 2 items, and then 3, and I doubted about this. It makes sense to convert it to an object, it's more clear.

**Comment by adrianq** [comment]
@tokland PR seems to be working fine. I have created #583 as we agreed. I think we should take care of this sooner rather than later as otherwise, it can be a bit frustrating for the user. Regarding https://github.com/EyeSeeTea/metadata-synchronization/issues/532, missing tasks are actually missing or you just forgot to tick them?

**Comment by tokland** [comment]
Cool. Ticks just were forgotten, updated. I did no test, we reused some use cases so it was not so helpful.

**Pull request merged**